python/langsmith/evaluation/_runner.py

```diff
@@ -376,38 +377,45 @@ def __init__(
     ):
         self._manager = experiment_manager
         self._results: List[ExperimentResultRow] = []
+        self._queue = queue.Queue()
         self._lock = threading.RLock()
-        self._thread = threading.Thread(
-            target=lambda: self._process_data(self._manager)
-        )
+        self._processing_complete = threading.Event()
+        self._thread = threading.Thread(target=self._process_data)
         self._thread.start()
 
     @property
     def experiment_name(self) -> str:
         return self._manager.experiment_name
 
     def __iter__(self) -> Iterator[ExperimentResultRow]:
-        processed_count = 0
         while True:
-            with self._lock:
-                if processed_count < len(self._results):
-                    yield self._results[processed_count]
-                    processed_count += 1
-                elif not self._thread.is_alive():
+            try:
+                item = self._queue.get(block=True, timeout=0.1)
+                if item is None:  # Sentinel value to indicate completion
+                    break
+                yield item
+            except queue.Empty:
+                if self._processing_complete.is_set():
                     break
 
-    def _process_data(self, manager: _ExperimentManager) -> None:
+    def _process_data(self) -> None:
         tqdm = _load_tqdm()
-        results = manager.get_results()
+        results = self._manager.get_results()
         for item in tqdm(results):
+            self._queue.put(item)
             with self._lock:
                 self._results.append(item)
-        summary_scores = manager.get_summary_scores()
+
+        summary_scores = self._manager.get_summary_scores()
         with self._lock:
             self._summary_results = summary_scores
 
+        self._queue.put(None)  # Sentinel value to indicate completion
+        self._processing_complete.set()
```

Comment:
sorry noob question as an outsider coming in 😆: is this None sentinel redundant with `_processing_complete`? Seems like they both just serve as a check for the same thing

Reply:
Great point! These can be collapsed into one condition -- made the fix.